[SPARK-25790][MLLIB] PCA: Support more than 65535 column matrix

shahidki31 opened the pull request:
## What changes were proposed in this pull request?
Spark PCA supports maximum only ~65,535 columns matrix. This is due to the fact that, it computes the Covariance matrix first, then compute principle components. The main bottle neck was computing **covariance matrix.** The limit 65,500 came due to the integer size limit. Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

Currently we don't have such limitation for computing SVD in spark.  So, we can make use of Spark SVD to compute the PCA, if the number of columns exceeds the limit.

Computation of PCA can be done directly using SVD of matrix, instead of finding the covariance matrix.
Following are the papers/links for the reference.

https://arxiv.org/pdf/1404.1100.pdf
https://en.wikipedia.org/wiki/Principal_component_analysis#Singular_value_decomposition
http://www.ifis.uni-luebeck.de/~moeller/Lectures/WS-16-17/Web-Mining-Agents/PCA-SVD.pdf

## How was this patch tested?
added UT, also manually verified with the existing test for pca, by removing the limit condition in the fit method.

shahidki31 committed:
[SPARK-25790][MLLIB] PCA: Support more than 65535 column matrix

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

shahidki31 commented:
Test results with existing PCA and using SVD without computing covariance matrix
val data = Array(
    Vectors.sparse(5, Seq((1, 1.0), (3, 7.0))),
    Vectors.dense(2.0, 0.0, 3.0, 4.0, 5.0),
    Vectors.dense(4.0, 0.0, 0.0, 6.0, 7.0))

1) PCA using covariance matrix
explained Variance = [ 0.7943932532, 0.2056067468, 1.26E-16]
Top 2 Principle components :  
[[-0.44859172075072673 -0.28423808214073987 
0.13301985745398526 -0.05621155904253121 
-0.1252315635978212 0.7636264774662965  
0.21650756651919933 -0.5652958773533949 
-0.8476512931126826 -0.11560340501314653 ]]

2) PCA using SVD, without computing covariance matrix: 
explained Variance = [0.7943932532, 0.2056067468, 5.55E-17]
Top 2 Principle components :  
[[-0.44859172075072673 -0.2842380821407399
0.13301985745398529 -0.056211559042531424
-0.12523156359782125 0.7636264774662964  
0.21650756651919945 -0.5652958773533953
-0.8476512931126826 -0.11560340501314664]]


**Leading Eigen Values MSE = 0.0
Leading eigen vectors MSE = 0.0**

shahidki31 commented:
cc @srowen

dongjoon-hyun commented:
ok to test.

SparkQA commented:
**[Test build #97673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97673/testReport)** for PR 22784 at commit [`1252526`](https://github.com/apache/spark/commit/12525266fe76b767974e5ba94cd131251bc7ed3e).

SparkQA commented:
**[Test build #97673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97673/testReport)** for PR 22784 at commit [`1252526`](https://github.com/apache/spark/commit/12525266fe76b767974e5ba94cd131251bc7ed3e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97673/
Test FAILed.

dongjoon-hyun commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala:
```diff
@@ -424,6 +424,28 @@ object Vectors {
   }
 
+  /**
+    * Returns the mean centered vector
```
@shahidki31 . Could you fix the scalastyle first in order to test Jenkins tests successfully?

shahidki31 replied:
Thank you @dongjoon-hyun . I have updated the code.

shahidki31 commented:
retest this please

SparkQA commented:
**[Test build #97677 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97677/testReport)** for PR 22784 at commit [`e6cf661`](https://github.com/apache/spark/commit/e6cf6612ee488ace7bfc11db26ee4d6cc72e3368).

SparkQA commented:
**[Test build #97678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97678/testReport)** for PR 22784 at commit [`9111fca`](https://github.com/apache/spark/commit/9111fcab296ca71bfa280010d60aa803ece69509).

SparkQA commented:
**[Test build #97678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97678/testReport)** for PR 22784 at commit [`9111fca`](https://github.com/apache/spark/commit/9111fcab296ca71bfa280010d60aa803ece69509).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97678/
Test FAILed.

SparkQA commented:
**[Test build #97679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97679/testReport)** for PR 22784 at commit [`4c1776f`](https://github.com/apache/spark/commit/4c1776f14f1453c2f64350a58cab7209764f826c).

SparkQA commented:
**[Test build #97677 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97677/testReport)** for PR 22784 at commit [`e6cf661`](https://github.com/apache/spark/commit/e6cf6612ee488ace7bfc11db26ee4d6cc72e3368).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97677/
Test PASSed.

SparkQA commented:
**[Test build #97680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97680/testReport)** for PR 22784 at commit [`9aff54f`](https://github.com/apache/spark/commit/9aff54fecc530c77e5f97941e15a478a421827d0).

SparkQA commented:
**[Test build #97679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97679/testReport)** for PR 22784 at commit [`4c1776f`](https://github.com/apache/spark/commit/4c1776f14f1453c2f64350a58cab7209764f826c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97679/
Test PASSed.

SparkQA commented:
**[Test build #97680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97680/testReport)** for PR 22784 at commit [`9aff54f`](https://github.com/apache/spark/commit/9aff54fecc530c77e5f97941e15a478a421827d0).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97680/
Test FAILed.

kiszk commented:
retest this please

SparkQA commented:
**[Test build #97682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97682/testReport)** for PR 22784 at commit [`9aff54f`](https://github.com/apache/spark/commit/9aff54fecc530c77e5f97941e15a478a421827d0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97682/
Test FAILed.

shahidki31 commented:
All the UTs are passing locally. Seems random error.
retest this please.

SparkQA commented:
**[Test build #97683 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97683/testReport)** for PR 22784 at commit [`23cfda1`](https://github.com/apache/spark/commit/23cfda1547355a823a3b2b2d374e64608c9ce175).

SparkQA commented:
**[Test build #97683 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97683/testReport)** for PR 22784 at commit [`23cfda1`](https://github.com/apache/spark/commit/23cfda1547355a823a3b2b2d374e64608c9ce175).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97683/
Test PASSed.

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does `65,500` come from?

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does `65,500` come from?

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does `65,500` come from?

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does this limitation `65,500` come from?

shahidki31 commented:
Hi @kiszk ,
 I think, INT_MAX is 2147483647, so n ~= sqrt(2*2147483647) = 65536.
Thanks

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does this limitation `65,500` come from?

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does this limitation `65,500` come from?

kiszk commented:
Can I clarify the description?
> Because we are passing an array of size n*(n+1)/2 to the breeze library and the size cannot be more than INT_MAX. so, the maximum column size we can give is 65,500.

If n > 20726, `n*(n+1)/2` > 214783647 ( = INT_MAX)`. Where does this limitation `65,500` come from?

kiszk commented:
One question: After this PR, what is the maximum column that we can accept?

shahidki31 commented:
Hi @kiszk Maximum it can go upto the following limit.
https://github.com/apache/spark/blob/23cfda1547355a823a3b2b2d374e64608c9ce175/mllib/src/main/scala/org/apache/spark/mllib/linalg/EigenValueDecomposition.scala#L78-L79

where ncv = min(n, 2*k), normally k << n.

For eg: if n = 1 million features, we can compute top 100 principle components. 

Number of principle components to compute is configurable.

shahidki31 commented:
Hi @srowen , Thanks for the comment. As per my knowledge, PCA/SVD is not limited on row size.

1) Currently row size is not a constraint. Ultimately we need to compute graminan matrix/GramianMatrix vector product to compute SVD.  So, computation of svd is limited by columns only.

5) Sparsity is only for the computation of gramian matrix/ gramian matrix vector product in both PCA and Spark SVD. Mean centred vector  will always be dense. Currently PCA is computed with dense matrix and SVD uses dense vector.  
So, only constraint about dense is coming in the matrix vector product computation.

6) In this PR, if the limit exceeds, it will compute in the distributed manner, which current PCA doesn't support.

2) Currently PCA is not scalable in terms of column number
 For 40GB driver memory, and number of columns is 40,000 and number of rows is 1lakh, I am getting following error.

```
scala> val pca = new PCA(k).fit(rad)
2018-10-22 22:44:23,128 | WARN  | main | 40000 columns will require at least 12800 megabytes of memory! | org.apache.spark.internal.Logging$class.logWarning(Logging.scala:66)
2018-10-22 22:47:02,836 | WARN  | main | 40000 columns will require at least 12800 megabytes of memory! | org.apache.spark.internal.Logging$class.logWarning(Logging.scala:66)
java.lang.OutOfMemoryError
  at java.io.ByteArrayOutputStream.hugeCapacity(ByteArrayOutputStream.java:123)
  at java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:117)
  at java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:93)
  at java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:153)
  at org.apache.spark.util.ByteBufferOutputStream.write(ByteBufferOutputStream.scala:41)
  at java.io.ObjectOutputStream$BlockDataOutputStream.drain(ObjectOutputStream.java:1877)
  at java.io.ObjectOutputStream$BlockDataOutputStream.setBlockDataMode(ObjectOutputStream.java:1786)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1189)
  at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
  at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:43)

```

kiszk commented:
Sorry for my mistake. My keyboard '4' sometimes has a trouble.
> I think, INT_MAX is 2147483647, so n ~= sqrt(2*2147483647) = 65536.

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -54,4 +55,14 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
     assert(PCAUtil.memoryCost(40000, 60000) > Int.MaxValue)
   }
+
+  test("number of features more than 65500") {
```
65535?

shahidki31 replied:
Updated. Thanks

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -54,4 +55,14 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
+    val randomRDD = RandomRDDs.normalVectorRDD(sc, rows, columns, 0, 0)
+    val pca = new PCA(k).fit(randomRDD)
+    assert(pca.explainedVariance.size === 5)
+    assert(pca.pc.numRows === 100000 && pca.pc.numCols === 5)
```
I wonder if there's any reasonable way to check the answer here, like some bounds on what the eigenvalues/eigenvectors should be? Like the eigenvalues would at least be positive?

shahidki31 replied:
Added two more assert, for checking positive eigen values and norm of principle components.

srowen replied:
Is there an easy dummy test case we can write where we know what the first PC should be? like if you generate a bunch of vectors like (a +/- epsilon, a +/- epsilon, ...) for many a, the principal component should be (1,1,1...) nearly right? is that easy enough to add as a trivial test of the actual analysis? I think that would really prove it, though you manual test suggests it's working.

shahidki31 replied:
Thanks. I have updated the test case.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:
```diff
@@ -384,18 +384,28 @@ class RowMatrix @Since("1.0.0") (
-    val Cov = computeCovariance().asBreeze.asInstanceOf[BDM[Double]]
+    if (n > 65535) {
+      val svd = computeSVD(k)
+      val s = svd.s.toArray.map(eigValue => eigValue * eigValue / (n - 1))
```
My linear algebra is probably rusty, so check me here, but we need the eigendecomposition of the covariance matrix 1/(n-1) * mat' * mat. I get how the singular values here give these eigenvalues. Don't the singular vectors V need to be divided by n-1 too? I'm probably wrong about it, just checking, esp. as we don't actually have tests for any of the output here!

shahidki31 replied:
Eigen vectors has unit norm. So scaling doesn't matter. So, we should scale only eigen value matrix.

Actually I have tested the PR against smaller matrix, which is shown in the comment https://github.com/apache/spark/pull/22784#issuecomment-431613820 , by removing the limit check

srowen replied:
Right, make sense.

srowen reviewed (commented):
OK I'm more convinced this is a useful improvement. Mostly interested in making sure it's correct now. I wonder if the SVD should be used at even smaller scales? as you point out, it's pretty hard to compute a gramian on even a 40k x 40k matrix.

SparkQA commented:
**[Test build #98134 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98134/testReport)** for PR 22784 at commit [`b1789d7`](https://github.com/apache/spark/commit/b1789d7a2305c53b463960e1d60f85abde5934ad).

shahidki31 commented:
Thank you @srowen for the review.  I have addressed the comments.

>  I wonder if the SVD should be used at even smaller scales? as you point out, it's pretty hard to compute a gramian on even a 40k x 40k matrix. 
> 

Yes. We can compute the PCA using SVD even for smaller scales. In fact if the number of columns are lesser, Spark SVD computes eigen decomposition by computing gramian matrix first, which is the same approach as in PCA.

The condition for whether to compute gramian matrix first or not is given below, for Spark SVD.
https://github.com/apache/spark/blob/d5573c578a1eea9ee04886d9df37c7178e67bb30/mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala#L232-L243




So, for smaller number of columns, (< 15000 columns), Spark SVD prefers computation of graminan matrix first and then computing the svd, which is same as the current implimentation of PCA.

shahidki31 committed:
address comments

SparkQA commented:
**[Test build #98135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98135/testReport)** for PR 22784 at commit [`a8c4391`](https://github.com/apache/spark/commit/a8c43919a5d8624a5a5ddf7ea862a93f2db098c6).

SparkQA commented:
**[Test build #98134 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98134/testReport)** for PR 22784 at commit [`b1789d7`](https://github.com/apache/spark/commit/b1789d7a2305c53b463960e1d60f85abde5934ad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98134/
Test PASSed.

SparkQA commented:
**[Test build #98135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98135/testReport)** for PR 22784 at commit [`a8c4391`](https://github.com/apache/spark/commit/a8c43919a5d8624a5a5ddf7ea862a93f2db098c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98135/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/feature/PCA.scala:
```diff
@@ -49,7 +50,16 @@ class PCA @Since("1.4.0") (@Since("1.4.0") val k: Int) {
 
-    val mat = new RowMatrix(sources)
+    val mat = if (numFeatures > 65535) {
+      val meanVector = Statistics.colStats(sources).mean
```
Rather than call `.toArray` and `.zipped` below, can this not be written as Vector - Vector in the loop below? might be more efficient.

shahidki31 replied:
I have modified. Thanks

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -54,4 +55,21 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
+    assert(pca.explainedVariance.size === 5)
+    assert(pca.pc.numRows === 100000 && pca.pc.numCols === 5)
+    // Eigen values should not be negative
+    assert(!pca.explainedVariance.values.exists(_ < 0))
```
You can write `.forAll(_ >= 0)` too, but doesn't matter

shahidki31 replied:
Done.

srowen commented on mllib/src/test/scala/org/apache/spark/mllib/feature/PCASuite.scala:
```diff
@@ -54,4 +55,21 @@ class PCASuite extends SparkFunSuite with MLlibTestSparkContext {
+    // Eigen values should not be negative
+    assert(!pca.explainedVariance.values.exists(_ < 0))
+
+    // Norm of the principle component should be 1.0
```
Nit: principle -> principal

shahidki31 replied:
Done.

SparkQA commented:
**[Test build #98140 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98140/testReport)** for PR 22784 at commit [`5674e17`](https://github.com/apache/spark/commit/5674e177b7894d61904c6748dbf7721359163938).

SparkQA commented:
**[Test build #98141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98141/testReport)** for PR 22784 at commit [`3cbe017`](https://github.com/apache/spark/commit/3cbe017c640764db0fe95bcc2a820917bbc5fb3e).

SparkQA commented:
**[Test build #98140 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98140/testReport)** for PR 22784 at commit [`5674e17`](https://github.com/apache/spark/commit/5674e177b7894d61904c6748dbf7721359163938).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98140/
Test PASSed.

SparkQA commented:
**[Test build #98141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98141/testReport)** for PR 22784 at commit [`3cbe017`](https://github.com/apache/spark/commit/3cbe017c640764db0fe95bcc2a820917bbc5fb3e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98141/
Test PASSed.

shahidki31 committed:
added test

SparkQA commented:
**[Test build #98144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98144/testReport)** for PR 22784 at commit [`094594b`](https://github.com/apache/spark/commit/094594bf63a22be65bac7b31932d5d870f1142d3).

SparkQA commented:
**[Test build #98145 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98145/testReport)** for PR 22784 at commit [`18af032`](https://github.com/apache/spark/commit/18af0325e95552a00983983224795e71f2e66204).

SparkQA commented:
**[Test build #98144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98144/testReport)** for PR 22784 at commit [`094594b`](https://github.com/apache/spark/commit/094594bf63a22be65bac7b31932d5d870f1142d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98144/
Test PASSed.

SparkQA commented:
**[Test build #98145 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98145/testReport)** for PR 22784 at commit [`18af032`](https://github.com/apache/spark/commit/18af0325e95552a00983983224795e71f2e66204).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98145/
Test PASSed.

SparkQA commented:
**[Test build #98161 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98161/testReport)** for PR 22784 at commit [`0d9eea8`](https://github.com/apache/spark/commit/0d9eea8fcffbdd72bdb8dd8b93de3ac9a782fc85).

shahidki31 committed:
update

SparkQA commented:
**[Test build #98164 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98164/testReport)** for PR 22784 at commit [`2b7ee7b`](https://github.com/apache/spark/commit/2b7ee7b0a6d2cbcc159826d8dbe286a4a144d463).

SparkQA commented:
**[Test build #98161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98161/testReport)** for PR 22784 at commit [`0d9eea8`](https://github.com/apache/spark/commit/0d9eea8fcffbdd72bdb8dd8b93de3ac9a782fc85).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98161/
Test PASSed.

SparkQA commented:
**[Test build #98164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98164/testReport)** for PR 22784 at commit [`2b7ee7b`](https://github.com/apache/spark/commit/2b7ee7b0a6d2cbcc159826d8dbe286a4a144d463).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98164/
Test PASSed.

srowen reviewed (commented):
That's looking good. Pinging some ML people I haven't heard from in a while like @MLnick or @sethah

shahidki31 committed:
address comment

SparkQA commented:
**[Test build #98178 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98178/testReport)** for PR 22784 at commit [`0effc85`](https://github.com/apache/spark/commit/0effc85ccfc831bcc4c469b4a4c1d8db26fab72e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98178/
Test PASSed.

shahidki31 commented:
Thanks a lot @srowen
